README.md
# pyconsul

API client for [HashiCorp's Consul](https://www.consul.io).

# Background

Main two motivations behind this project:

- Create a Python API client for Consul that returned typed responses via
  models (Pydantic model objects or at least Python data classes).
- Learn about creating and maintaining an OSS project.
- Create a client library that supported OpenTelemetry.

At work we make use of the entire HashiCorp stack (Consul, Nomad, Vault, etc) and
for some of our projects, we're split between Go and Python. Most of the Python
clients that exist today didn't seem to fit the need that we had. Mainly that we are
striving to be more strongly typed to ease the development of either small scripts
or larger projects.

That and we've heavily invested into OpenTelemetry, which is something that seems
to be lacking in a number of projects. If it wasn't for the desire that I had
to return typed objects, I'd likely have looked to contribute to the other
existing projects.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.3.1] - 2023-01-13

#### Fixed

- Ensuring `*_ipv4` TaggedAddresses are accounted for.
  - Proper casing was adjusted for `address` and `port` TaggedAddress attributes.

## [0.3.0] - 2023-01-13

### Added

- A little context awareness when working with services running in Nomad.
  - Example, `ServiceHealth` now has the `alloc_id` property

### Changed

- Limiting usage of imports from `typing`
  - For example, moving more hinting of lists from `List` to `list`

### Internal

- Added `Makefile` to make a few things easier to call all the time.

## [0.2.0] - 2022-10-27

### Added

- `/health/service/:service` support with model based return
- Tracing decorator for each endpoint call for basic span generation
- `timeout` option to the base `ConsulAPI` class to ensure we have a default timeout
  set but allow it to be overriden.

## [0.1.0] - 2022-10-24

### Changed

- Switched how we parse env vars for the following:
  - `CONSUL_HTTP_ADDR`
  - `CONSUL_HTTT_TOKEN`
  - `CONSUL_NAMESPACE`
  -

### Added

- `/catalog/services` support with a basic example




tests/test_lib_consul.py
import pytest
import respx

from hc_pyconsul.lib.consul import ConsulAPI
from hc_pyconsul.lib.exceptions import Unauthenticated


@respx.mock
def test_get_403(respx_mock):
    nomad = ConsulAPI()
    endpoint = '/testing403'
    mocked_route = respx_mock.get(url=f'{nomad.address}/v1{endpoint}')
    mocked_route.respond(status_code=403, json={})

    with pytest.raises(Unauthenticated):
        nomad._get(endpoint=endpoint)


tests/test_helpers_services.py
import unittest

from hc_pyconsul.helpers.services import extract_alloc_id_from_service_name


class TestHelpersExtractAllocationID(unittest.TestCase):
    def test_extract_alloc_id(self):
        results = extract_alloc_id_from_service_name(
            service_id='_nomad-task-86c6736f-33d2-e3b1-69a2-8189031eb599-service-01'
        )

        assert results.id == '86c6736f'
        assert results.id_long == '86c6736f-33d2-e3b1-69a2-8189031eb599'

    def test_extract_alloc_id_unknown_pattern(self):
        results = extract_alloc_id_from_service_name(
            service_id='86c6736f-33d2-e3b1-69a2-8189031eb599'
        )

        assert results.id == ''
        assert results.id_long == ''


tests/test_lib_consul_health.py
from unittest import TestCase

import pytest
import respx
from httpx import Response

from hc_pyconsul import ConsulHealth
from hc_pyconsul.exceptions import Unauthenticated
from hc_pyconsul.lib.consul import ConsulAPI
from hc_pyconsul.models.health import ServiceHealth
from tests.fixtures.health import responses


@respx.mock
def test_get_403(respx_mock):
    nomad = ConsulAPI()
    endpoint = '/testing403'
    mocked_route = respx_mock.get(url=f'{nomad.address}/v1{endpoint}')
    mocked_route.respond(status_code=403, json={})

    with pytest.raises(Unauthenticated):
        nomad.call_api(endpoint=endpoint, verb='GET')


class TestListServiceInstances(TestCase):

    def setUp(self) -> None:
        self.test_health = ConsulHealth()

    def tearDown(self) -> None:
        self.test_health = None

    @respx.mock
    def test_valid_response(self):
        response = respx.get('http://localhost:8500/v1/health/service/foobar')
        response.return_value = Response(200, json=responses.HEALTH_SERVICES_LIST)

        services: list[ServiceHealth] = self.test_health.list_service_instances(service='foobar')

        self.assertEqual('redis', services[0].service.id)

    @respx.mock
    def test_valid_response_tagged_addresses(self):
        response = respx.get('http://localhost:8500/v1/health/service/foobar')
        response.return_value = Response(200, json=responses.LAN_WAN_IPV4_TAGGED_ATTRIBUTES)

        services: list[ServiceHealth] = self.test_health.list_service_instances(service='foobar')

        self.assertEqual('10.1.10.12', services[0].service.tagged_addresses.lan.address)
        self.assertEqual('198.18.1.2', services[0].service.tagged_addresses.wan.address)

    @respx.mock
    def test_valid_response_with_nomad_service(self):
        response = respx.get('http://localhost:8500/v1/health/service/foobar')
        response.return_value = Response(200, json=responses.SERVICE_LIST_NOMAD)

        services: list[ServiceHealth] = self.test_health.list_service_instances(service='foobar')

        self.assertEqual('86c6736f', services[0].alloc_id.id)
        self.assertEqual('86c6736f-33d2-e3b1-69a2-8189031eb599', services[0].alloc_id.id_long)


hc_pyconsul/lib/catalog.py
from typing import Dict
from typing import List

from pydantic.dataclasses import dataclass

from hc_pyconsul.lib.consul import ConsulAPI


@dataclass
class ConsulCatalog(ConsulAPI):

    # pylint: disable=invalid-name
    def list_services(self, dc: str = None, node_meta: List[str] = None, namespace: str = None) -> Dict[str, List[str]]:
        """
        Link to official docs:
            https://developer.hashicorp.com/consul/api-docs/catalog#list-services

        Parameters
        ----------
        dc: str = None
            Specifies the datacenter to query.
            This will default to the datacenter of the agent being queried.
        node_meta: List[str] = None
            Specifies a desired node metadata key/value in the form of key:value.
            This parameter can be specified multiple times,
            and filters the results to nodes with the specified key/value pairs.
        namespace: str = None
            Enterprise Only
            Specifies the namespace of the services you lookup.

        Returns
        -------
        services: Dict[str, List[str]]

        """
        params = {}
        if dc:
            params.update({
                'dc': dc
            })

        if node_meta:
            params.update({
                'node-meta': ','.join(meta_item for meta_item in node_meta)
            })

        if namespace:
            params.update({
                'ns': namespace
            })

        results = self._get(endpoint='/catalog/services', params=params)

        return results


hc_pyconsul/lib/consul.py
import os
from typing import Optional

import httpx
from httpx import HTTPStatusError
from pydantic.dataclasses import dataclass

from hc_pyconsul.lib.exceptions import Unauthenticated
from hc_pyconsul.lib.exceptions import UnknownResourceCalled


@dataclass
class ConsulAPI:
    address: str = "http://localhost:8500"
    token: Optional[str] = None
    namespace: Optional[str] = None

    def __post_init__(self):

        if os.environ.get('CONSUL_HTTP_ADDR'):
            self.address = os.environ.get('CONSUL_HTTP_ADDR')

        if os.environ.get('CONSUL_HTTP_TOKEN'):
            self.token = os.environ.get('CONSUL_HTTP_TOKEN')

        if os.environ.get('CONSUL_NAMESPACE'):
            self.namespace = os.environ.get('CONSUL_NAMESPACE')

    def _get(self, endpoint: str, **kwargs):
        """
        Specifically making GET requests to Consul that expect JSON.

        Parameters
        ----------
        endpoint: str
            The specific endpoint to hit.

        Returns
        -------
        response

        Raises
        ------
        Unauthenticated
        UnknownResourceCalled
        """

        call_headers = kwargs.get('headers', {})

        if self.token:
            call_headers.update(
                {
                    'X-Consul-Token': self.token
                }
            )

        if self.namespace:
            if not kwargs.get('params'):
                kwargs['params'] = {}

            kwargs['params'].update({'namespace': self.namespace})

        url = f'{self.address}/v1{endpoint}'
        try:

            response = httpx.get(url, headers=call_headers, **kwargs)

            response.raise_for_status()

        # Attempt to gracefully handle Consul's documented response codes with helpful exceptions
        except HTTPStatusError as http_error:
            if http_error.response.status_code == 403:
                raise Unauthenticated('API called failed to due being unauthenticated. Maybe your token expired?')
            if http_error.response.status_code == 404:
                raise UnknownResourceCalled('Received a 404. Check supported API endpoints for your version of Consul.')
            raise http_error

        return response.json()
